### PR TITLE
Create IR structs, and merge functions for IRs.

### LINKED ISSUE #125: Support Gateway API extensions (output providers)
**What would you like to be added**:

Add the ability to translate `provider` specific APIs into implementation specific Gateway API extensions (output providers) for fields not currently supported in the Gateway API

**Why this is needed**:
Adding ability to translate all input provider specific intent without losing any functionality
For e.g. the current [ingress-nginx](https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/ingressnginx/README.md#ingress-nginx-provider) translation only translates a subset of the API (annotations)  https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#annotations 
Without this feature, its unlikely that the user will migrate to Gateway API until all the input provider features are supported in the Gateway API APIs
If this tool supports the notion of output providers, the input API could be translated to various implementation specific Gateway API extensions

### PR DESCRIPTION
**Pull request by sawsa307**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add structures needed to support translation of extension features.

**Which issue(s) this PR fixes**:

Fixes #125

**Does this PR introduce a user-facing change?**:

```release-note
None
```

### CODE REVIEW

File: pkg/i2gw/provider.go [hunk 1]
```diff
@@ -101,6 +102,72 @@ type GatewayResources struct {
...
+// ServiceIR contains a dedicated field for each provider to specify their
+// extension features on Service.
+type ServiceIR struct {
+	Apisix       *ir.ApisixServiceIR
+	Gce          *ir.GceServiceIR
+	IngressNginx *ir.IngressNginxServiceIR
+	Istio        *ir.IstioServiceIR
+	Kong         *ir.KongServiceIR
+	Openapi3     *ir.Openapi3ServiceIR
+}
```
Comment (LiorLieberman) [inline pkg/i2gw/provider.go:169]: is there a reason why all these new structs are not part of the IR package? Not saying it should move, just want to get your thoughts and see if it does/doesn't make sense

File: pkg/i2gw/provider.go [hunk 2]
```diff
@@ -101,6 +102,72 @@ type GatewayResources struct {
...
+// has a dedicated field for each provider to specify their extension features
+// on Gateways.
+// The IR will contain necessary information to construct the Gateway
+// extensions, but not the extensions themselves.
+type GatewayContext struct {
+	gatewayv1.Gateway
+	GatewayIR
+}
+
+type GatewayIR struct {
```
Comment (LiorLieberman) [inline pkg/i2gw/provider.go:132]:
```go
type GatewayContext struct {
	gatewayv1.Gateway
	providerSpecificIR ProviderSpecificGatewayIR
}

type ProviderSpecificGatewayIR struct {
```

Comment (LiorLieberman) [reply pkg/i2gw/provider.go:132]: similar stuff for http and service

File: pkg/i2gw/ingress2gateway.go [hunk 3]
```diff
@@ -198,3 +199,74 @@ func mergeGateways(gatewaResources []GatewayResources) (map[types.NamespacedName
...
+		ReferenceGrants: make(map[types.NamespacedName]gatewayv1beta1.ReferenceGrant),
+	}
+	var errs field.ErrorList
+	mergedIRs.Gateways, errs = mergeGatewayContexts(irs)
+	if len(errs) > 0 {
+		return IR{}, errs
+	}
+	for _, gr := range irs {
+		maps.Copy(mergedIRs.GatewayClasses, gr.GatewayClasses)
+		maps.Copy(mergedIRs.HTTPRoutes, gr.HTTPRoutes)
```
Comment (LiorLieberman) [inline pkg/i2gw/ingress2gateway.go:230]: I re-added the comment on merging httpIRs and ServiceIRs here https://github.com/kubernetes-sigs/ingress2gateway/pull/185#discussion_r1745843418

Please either add logic to merge them or add a clear TODO plus an i2gw issue to implement that

File: pkg/i2gw/ingress2gateway.go [hunk 4]
```diff
@@ -198,3 +199,74 @@ func mergeGateways(gatewaResources []GatewayResources) (map[types.NamespacedName
...
+			if len(g.Spec.Addresses) > 16 {
+				fieldPath := field.NewPath(fmt.Sprintf("%s/%s", nn.Namespace, nn.Name)).Child("spec").Child("addresses")
+				errs = append(errs, field.Invalid(fieldPath, g, "error while merging gateway listeners: a gateway cannot have more than 16 addresses"))
+			}
+		}
+	}
+	return newGatewayContexts, errs
+}
+
+func mergedGatewayIR(current, existing GatewayIR) GatewayIR {
```
Comment (LiorLieberman) [inline pkg/i2gw/ingress2gateway.go:267]: I dont like the specific need to call merge here... but I understand its a pain to solve now. Please add a todo and an issue on the repo so we can get this PR merged but we should revisit this soon.

**Review by LiorLieberman — COMMENTED** [review]
I appreciate you put the time to split this to two more concise commits. This is very useful, thanks @sawsa307 !

File: pkg/i2gw/provider.go (hunk 2 above)
Comment (sawsa307) [reply pkg/i2gw/provider.go:132]: Done.

File: pkg/i2gw/provider.go (hunk 1 above)
Comment (sawsa307) [reply pkg/i2gw/provider.go:169]: I think it makes perfect sense to move them into the ir package. I didn't think of doing it previously.

Just created a new file called intermediate_representation.go to hold these structs.

File: pkg/i2gw/ingress2gateway.go (hunk 4 above)
Comment (sawsa307) [reply pkg/i2gw/ingress2gateway.go]: Done, added an issue to track this

File: pkg/i2gw/ingress2gateway.go (hunk 3 above)
Comment (sawsa307) [reply pkg/i2gw/ingress2gateway.go]: Done. Added an issue to track this.

**Review by LiorLieberman — APPROVED** [review]
This Looks good to me, I also tested it locally with two files. 
Thank you very much for all your hard work @sawsa307 💪!

/approve
lets get final lgtm from @robscott for the specific logic

one is:
`./ingress2gateway print --input-file tmp.yaml -A --providers gce`

tmp.yaml
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: sa-test
  annotations:
    kubernetes.io/ingress.class: "gce"
spec:
  rules:
  - http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: whereami
            port:
              number: 80
---
apiVersion: cloud.google.com/v1
kind: BackendConfig
metadata:
  name: sa-test
spec:
  sessionAffinity:
    affinityType: "CLIENT_IP"
---
apiVersion: v1
kind: Service
metadata:
  name: whereami
  annotations:
    cloud.google.com/neg: '{"ingress": true}'
    cloud.google.com/backend-config: '{"default": "sa-test"}'
spec:
  selector:
    app: whereami
  ports:
    - port: 80
      protocol: TCP
      targetPort: 8080
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: whereami
spec:
  replicas: 3
  selector:
    matchLabels:
      app: whereami
  template:
    metadata:
      labels:
        app: whereami
    spec:
      containers:
      - name: whereami
        image: us-docker.pkg.dev/google-samples/containers/gke/whereami:v1.2.20
        ports:
          - name: http
            containerPort: 8080
        readinessProbe:
          httpGet:
            path: /healthz
            port: 8080
            scheme: HTTP
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deployment
spec:
  selector:
    matchLabels:
      purpose: bsc-config-demo
  replicas: 2
  template:
    metadata:
      labels:
        purpose: bsc-config-demo
    spec:
      containers:
      - name: hello-app-container
        image: us-docker.pkg.dev/google-samples/containers/gke/hello-app:1.0
```
output:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  creationTimestamp: null
  name: gce
spec:
  gatewayClassName: gke-l7-global-external-managed
  listeners:
  - name: http
    port: 80
    protocol: HTTP
status: {}
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  creationTimestamp: null
  name: sa-test-all-hosts
spec:
  parentRefs:
  - name: gce
  rules:
  - backendRefs:
    - name: whereami
      port: 80
    matches:
    - path:
        type: PathPrefix
        value: /
status:
  parents: []
---
apiVersion: networking.gke.io/v1
kind: GCPBackendPolicy
metadata:
  creationTimestamp: null
  name: whereami-GCPBackendPolicy
spec:
  default:
    sessionAffinity:
      type: CLIENT_IP
  targetRef:
    group: ""
    kind: Service
    name: whereami
status: {}
```

Second test to previous capabilities not break:
`./ingress2gateway print --input-file ingress.yaml -A --providers ingress-nginx`

ingress.yaml
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-ingress
spec:
  ingressClassName: nginx
  tls:
  - hosts:
    - foo.example.com
    - bar.example.com
    secretName: example-com
  rules:
  - host: foo.example.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: foo-app
            port:
              number: 80
      - path: /orders
        pathType: Prefix
        backend:
          service:
            name: foo-orders-app
            port:
              number: 80
  - host: bar.example.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: bar-app
            port:
              number: 80
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-canary-ingress
  annotations:
    nginx.ingress.kubernetes.io/canary: "true"
    nginx.ingress.kubernetes.io/canary-weight: "50"
spec:
  ingressClassName: nginx
  tls:
  - hosts:
    - foo.example.com
    - bar.example.com
    secretName: example-com
  rules:
  - host: foo.example.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: bar-app
            port:
              number: 80
```
output:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  creationTimestamp: null
  name: nginx
spec:
  gatewayClassName: nginx
  listeners:
  - hostname: bar.example.com
    name: bar-example-com-http
    port: 80
    protocol: HTTP
  - hostname: bar.example.com
    name: bar-example-com-https
    port: 443
    protocol: HTTPS
    tls:
      certificateRefs:
      - group: null
        kind: null
        name: example-com
  - hostname: foo.example.com
    name: foo-example-com-http
    port: 80
    protocol: HTTP
  - hostname: foo.example.com
    name: foo-example-com-https
    port: 443
    protocol: HTTPS
    tls:
      certificateRefs:
      - group: null
        kind: null
        name: example-com
      - group: null
        kind: null
        name: example-com
status: {}
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  creationTimestamp: null
  name: example-ingress-bar-example-com
spec:
  hostnames:
  - bar.example.com
  parentRefs:
  - name: nginx
  rules:
  - backendRefs:
    - name: bar-app
      port: 80
    matches:
    - path:
        type: PathPrefix
        value: /
status:
  parents: []
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  creationTimestamp: null
  name: example-canary-ingress-foo-example-com
spec:
  hostnames:
  - foo.example.com
  parentRefs:
  - name: nginx
  rules:
  - backendRefs:
    - name: bar-app
      port: 80
      weight: 50
    - name: foo-app
      port: 80
      weight: 50
    matches:
    - path:
        type: PathPrefix
        value: /
  - backendRefs:
    - name: foo-orders-app
      port: 80
    matches:
    - path:
        type: PathPrefix
        value: /orders
status:
  parents: []
```

**Comment by robscott** [comment]
Thanks for all the work on this @sawsa307 and also to @LiorLieberman for the great reviews!

/lgtm
/approve

(For future readers, most of my review for this work was actually on #185.)

**Pull request merged**